File: 10.py
#!/usr/bin/env python3
from sys import stdin
from statistics import median

pos, neg = '([{<', ')]}>'
ps1, ps2 = {')':3,']':57,'}':1197,'>':25137}, {')':1,']':2,'}':3,'>':4}

def P (xs,es=[]):
    if len(xs)==0: return (None, es)
    e, x, c = None if not es else es[0], xs[0], neg[pos.find(xs[0])]
    if e is None: return x if x in neg else P(xs[1:],[c])
    if x in pos:  return P(xs[1:], [c]+es)
    elif x != e:  return (x,es)
    else:         return P(xs[1:], es[1:])

scr = lambda xs, s=0: s if not xs else scr(xs[1:], 5*s+ps2.get(xs[0]))
ls  = list(map(str.strip, stdin.readlines()))
p1  = sum(map(lambda x: ps1.get(x[0]), filter(lambda x: x[0], list(map(P, ls)))))
p2  = median(map(lambda x: scr(x[1]), filter(lambda x: not x[0], map(P, ls))))

print(f"Part 1: { p1 }, Part 2: { p2 }")

File: 10.scala
#!/usr/bin/env scala
import scala.io.Source.stdin

@main def aoc(): Unit =
  val pos = "([{<"
  val neg = ")]}>"
  val ps1 = Map(')'->3, ']'->57, '}'->1197, '>'->25137)
  val ps2 = Map(')'->1, ']'->2, '}'->3, '>'->4)
  val ls  = stdin.getLines().toList

  def parse(xs:List[Char], es:List[Char]=List.empty): (Option[Char], List[Char]) = 
    val i = if xs.nonEmpty then pos.indexOf(xs(0)) else -1
    val n = if i == -1 then List() else List(neg(i))
    xs match
      case x :: t if es.isEmpty => parse(t, n)
      case x :: t if pos.contains(x) => parse(t, n ++ es)
      case x :: t if es.nonEmpty && x != es(0) => (Some(x), es)
      case x :: t => parse(t, es.tail)
      case _ => (None, es)

  val p1    = ls.flatMap{ x=>parse(x.toList)._1 }.collect(ps1(_)).sum
  val score = (xs:List[Char]) => xs.map(ps2(_).toLong).foldLeft(0L){ (a,x)=> a*5+x }
  val p2    = ls.map{ x=>parse(x.toList) }.collect { case (None, x) => score(x) }.sorted

  println(s"Part 1: ${ p1 }, Part 2: ${ p2(p2.length / 2)  }")
